pkg/pushsync/pushsync.go

```diff
@@ -77,54 +81,40 @@ func (ps *PushSync) Close() error {
 	return nil
 }
 
-// handler handles chunk delivery from other node and inserts it to localstore.
-// it also sends this chunk to the closest peer if one exists.
+// handler handles chunk delivery from other node and inserts it in to the local store,
+// It also sends a receipt for the chunk on the same stream.
+// ALso it forwards the received chunk to its closest peer.
 func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
-	_, r := protobuf.NewWriterAndReader(stream)
-	defer stream.Close()
-
-	var ch pb.Delivery
-
-	if err := r.ReadMsg(&ch); err != nil {
-		if err == io.EOF {
-			return nil
-		}
-		return err
-	}
-
-	// create chunk and store it in the local store
-	addr := swarm.NewAddress(ch.Address)
-	chunk := swarm.NewChunk(addr, ch.Data)
-	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+	chunk, err := ps.ReceiveChunkAndSendReceipt(ctx, stream)
```

Comment:
This is not correct forwarding. This handler will send receipt when it receives a delivery, but it should only if the peer that it forwards it to responds with its receipt, not before.

Reply:
1) I get a delivery
2) i forward to closest peer
3) closest peer forwards to closest peer

final peer put in DB and send a receipt which travels back...
when you get a receipt from your closest peer.. then put in your DB and send receipt to your predecessor?

Reply:
This implementation does not do that. It sends the receipt before the peer that it is forwarded to sends back the receipt.

I have other comments on the current design of this package, but this is the most crucial point that is not correct.

Reply:
Thanks for the comment @janos. 

Where is this behavior you said documented? Swarm book section 2.3.2 says that once storer node (which is closest to the chunks address) receives this, it should store it and send back the receipt.  Infact the book doesn't talk about forwarding the chunk from there on. The forwarding given in the book is  about the route the chunk follows to reach the closest peer (to the chunk address) after kademlia is implemented. Is my understanding right @zelig

Reply:
I am saying that this code does not do what you described. Storer node does not send the receipt.

Reply:
i am quoting again... Please read swarm book section 2.3.2.

Reply:
@jmozah strange. the second line of your comment above this function correctly describes what this funtion is doing:

>  // it sends a receipt for the chunk on the same stream.

this part is implemented as `ReceiveChunkAndSendReceipt` as the name suggests is what (only) the closest  peer should do.

> // ALso it forwards the received chunk to its closest peer.

Reply:
@zelig  So ReceiveChunkAndSendReceipt should wait to send receipt ..... only after sending the chunk to its closest peer and getting a receipt? 

Which of this 2 Algo is right?

Algo1:
  - Get a chunk delivery
  - add in DB
  - send a receipt in the same stream 
  - Find closest peer
  - Forward the received chunk  to the closest peer
   - receive receipt


Algo 2:
    - Get a chunk delivery
    - Find closest peer
    - Forward the received chunk  to the closest peer
     - receive receipt from closest peer
     - only then add chunk in localDB
     - send a receipt to original sender in the same stream

Reply:
@zelig @janos @acud  I hae modified the code to follow Algo 2. Please review it and let me know your comments.